File: pset6/readability/readability.py
from cs50 import get_string


def main():
    # ask user for a text
    text = get_string("Text: ").lower()

    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # Coleman-Liau index: index = 0.0588 * L - 0.296 * S - 15.8
    L = letters / words * 100
    S = sentences / words * 100

    index = round(0.0588 * L - 0.296 * S - 15.8)

    if index >= 16:
        print("Grade 16+")

    elif index < 1:
        print("Before Grade 1")

    else:
        print(f"Grade {round(index)}")


def count_letters(string):
    count = 0

    # count only the letters in the string
    for i in string:
        if i.isalpha():
            count += 1
    return count


def count_words(string):
    # split the string using a white space as separator, and count the items in the returned list
    return len(string.split(' '))


def count_sentences(string):
    # consider senteces as any sequence of characters ended with a ".", "1", or "?"
    count = 0
    for i in string:
        if i in ['.', '!', '?']:
            count += 1

    return count


main()

File: pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <ctype.h>
#include <string.h>
#include <math.h>

int count_letters(string s);
int count_words(string s);
int count_sentences(string s);


int main(void)
{
    // Get the text from the user
    string text = get_string("Text: ");

    // Convert the text to lowercase
    for (int i = 0; i < strlen(text); i++)
    {
        text[i] = tolower(text[i]);
    }

    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Calculates averages for the formula
    float L = (float) letters / (float) words * 100;
    float S = (float) sentences / (float) words * 100;

    // Calculates the Coleman–Liau index
    int index = round((0.0588 * L) - (0.296 * S) - 15.8);

    // Print the results
    if (index <= 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}


int count_letters(string s)
{
    // iterates trough the given string and count only the letters
    int letters = 0;
    for (int i = 0; i < strlen(s); i++)
    {
        if (s[i] >= 'a' && s[i] <= 'z')
        {
            letters++;
        }
    }
    return letters;
}

int count_words(string s)
{
    // Counts words in the given string based on spaces
    int words = 1;
    for (int i = 0; i < strlen(s); i++)
    {
        if (s[i] == ' ')
        {
            words++;
        }
    }
    return words;
}

int count_sentences(string s)
{
    // Counts sentences in given string, using a . or a ! or a ? as separators
    int sentences = 0;
    for (int i = 0; i < strlen(s); i++)
    {
        if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            sentences++;
        }
    }
    return sentences;
}
